asmcomp/amd64/emit.mlp

```diff
@@ -869,10 +872,15 @@ let emit_instr env fallthrough i =
           emit_call "caml_reraise_exn";
           record_frame env Reg.Set.empty (Dbg_raise i.dbg)
       | Lambda.Raise_notrace ->
-          I.mov (domain_field Domainstate.Domain_exn_handler) rsp;
-          I.pop (domain_field Domainstate.Domain_exn_handler);
-          I.pop r11;
-          I.jmp r11
+          if Config.tsan then begin
+            emit_call "caml_tsan_raise_notrace_exn";
+            record_frame env Reg.Set.empty (Dbg_raise i.dbg)
```

Comment:
After taking a closer look, the current code seems correct to us:

`raise_notrace` will never stash the backtrace, and so normally does not use frame descriptors. But, if `tsan` is enabled, it will call a C function in `runtime/tsan.c` to update TSan’s view of the stack. That C function unwinds the stack up to the exception handler using frame descriptors. In order to be able to do that, the return address for the function called here, `caml_tsan_raise_notrace_exn`, must be recorded in a frame descriptor as well.